adr: lib2p implementation plan (#7282)
tendermint/tendermint: docs/architecture/adr-073-libp2p.md

# ADR 073: Adopt LibP2P

## Changelog

- 2021-11-02: Initial Draft (@tychoish)

## Status

Proposed.

## Context

As part of the 0.35 development cycle, the Tendermint team completed
first phase of the work described in ADRs 61 and 62, which included a
large scale refactoring of the reactors and the p2p message
routing. This replaced the switch and many of the other legacy
components, without breaking protocol or network-level
interoperability and left the legacy connection/socket handling code.

> Following the release, the team has reexamined the state of the code
> and the design, as well as Tendermints requirements for. The notes

Reviewer: ```suggestion
Following the release, the team has reexamined the state of the code
and the design, as well as Tendermint's requirements. The notes
```

> from this process are available in the [P2P Roadmap

Reviewer: ```suggestion
from that process are available in the [P2P Roadmap
```

Minor nit only to distinguish "this" meaning this document from "this" meaning the previous document referenced.

RFC](../rfc/rfc-000-p2p.rst).

> While this ADR supersedes the decisions made in ADRs 60 and 61, it is
> also enabled by the completed portions of this work. Previously, the

Reviewer: Utterly optional suggestion.

```suggestion
This ADR supersedes the decisions made in ADRs 60 and 61, but
builds on the completed portions of this work. Previously, the
```

boundaries of peer management, message handling, and the higher level

> business logic (e.g. "the reactors") were intermingled, and core

Reviewer: ```suggestion
business logic (e.g., "the reactors") were intermingled, and core
```

elements of the p2p system were responsible for the orchestration of

> the higher level business logic. The completed aspects of the refactor
> made it more obvious that the design implications of the legacy
> components had outsized influence on the entire implementation, that
> it would be difficult to iterate within the current abstractions, and
> it would not be viable to maintain interoperability with legacy
> systems while also achieving any of our broader objectives.

Reviewer: Suggestions to break up the long sentence a bit.

```suggestion
higher-level business logic. Refactoring the legacy components
made it more obvious that this entanglement of responsibilities
had outsized influence on the entire implementation, making
it difficult to iterate within the current abstractions.
It would not be viable to maintain interoperability with legacy
systems while also achieving many of our broader objectives.
```

> LibP2P is a thoroughly specificed implementation of a peer-to-peer

Reviewer: ```suggestion
LibP2P is a thoroughly-specified implementation of a peer-to-peer
```

networking stack, designed specifically for systems such as
ours. Adopting LibP2P as the basis of the Tendermint will allow the
Tendermint team to focus more of their time on other differentiating
aspects of the system, and make it possible for the ecosystem as a
whole to take advantage of tooling and efforts of the LibP2P
platform.

## Alternative Approaches

> As discussed in the related RFC, the primary alternative is to

Reviewer: ```suggestion
As discussed in the [P2P Roadmap RFC][rfc], the primary alternative would be to
```

> continue development of tendermint's home grown peer-to-peer

Reviewer: ```suggestion
continue development of Tendermint's home-grown peer-to-peer
```

> layer. While this gives the Tendermint team a maximal level of control

Reviewer: ```suggestion
layer. While that would give the Tendermint team maximal control
```

> over the peer system, the current peer system is unexceptional on its

Reviewer: ```suggestion
over the peer system, the current design is unexceptional on its
```

own merits, and the prospective maintenance burden for this system
exceeds our tolerances for the medium term.

> It is also the case that Tendermint can and should differentiate
> itself not on the basis of its networking implementation or peer
> management tools, but on its consistent operator experience,
> battletested consensus algorithm, and ergonomic user experience.

Reviewer: ```suggestion
Tendermint can and should differentiate itself not on the basis of
its networking implementation or peer management tools, but providing
a consistent operator experience, a battle-tested consensus algorithm,
and an ergonomic user experience.
```

## Decision

Tendermint will adopt libp2p during the 0.37 development cycle,
replacing the bespoke Tendermint P2P stack. This will remove the
`Endpoint`, `Transport`, `Connection`, and `PeerManager` abstractions
and leave the reactors, `p2p.Router` and `p2p.Channel`
abstractions.

LibP2P may obviate the need for a dedicated peer exchange (PEX)
reactor, which would also in turn obviate the need for a dedicated
seed mode. If this is the case, then all of this functionality would
be removed.

> If it turns out, based on the advice of Protocol Labs, that it makes

Reviewer: ```suggestion
If it turns out (based on the advice of Protocol Labs) that it makes
```

sense to maintain separate pubsub or gossipsub topics

> per-message-type, then the `Router` abstraction may dissolve
> entirely.

Reviewer: ```suggestion
per-message-type, then the `Router` abstraction could also 
be entirely subsumed.
```

## Detailed Design

> ### Implementation Changes

Reviewer 2: will we kill our homecooked handshake protocol in favour of noise? 

This would close a couple security issues we have with our secret handshake.

Author: Yes, the handshake protocol will go away (and I think is implied by "replacing the connection," though I can call it out specifically.

Reviewer 3: Is there a specific secure connection/identity mechanism in libp2p that can easily replace our handshake?

Reviewer: I believe by default libp2p uses conventional TLS handshakes, at least for TCP. I'm less clear about QUIC, but I think it might also apply there. There are hooks for a custom handshake, though.

> 

Reviewer 2: Creating a second thread for another question. Has there been a discussion for tcp vs quic? I assume we would use tcp for the first version?

Author: I will put this in open questions and/or discussion. 

I assume we'll default to tcp, but allow quic as an option.

> The "seams" in the P2P implementation between the higher level
> constructs (reactors), the routing layer (`Router`) and the lower
> level connection and peer management code make this operation
> relatively straight forward from an implementation perspective. The
> goal in this design is to minimize the impact to the reactors
> (potentially entirely,) and completely remove the lower level
> aspects (e.g. `Transport`, `Connection` and `PeerManager`) using the
> separation afforded by the `*Router` layer. The current state of the
> code makes these changes relatively surgical, and limited to a small
> number of methods:

Reviewer: ```suggestion
The seams in the P2P implementation between the higher level
constructs (reactors), the routing layer (`Router`) and the lower
level connection and peer management code make this operation
relatively straightforward to implement. A key
goal in this design is to minimize the impact on the reactors
(potentially entirely,) and completely remove the lower level
components (e.g., `Transport`, `Connection` and `PeerManager`) using the
separation afforded by the `*Router` layer. The current state of the
code makes these changes relatively surgical, and limited to a small
number of methods:
```

- `p2p.Router.OpenChannel` will still return a `Channel` structure
  which will continue to serve as a pipe between the reactors and the
  `Router`. The implementation will no longer need the queue

>   implementation and this operation will instead start goroutines that
>   are responsible for routing the messages from the channel to libp2p
>   fundamentals, replacing the current `p2p.Router.routeChannel`.

Reviewer: ```suggestion
  implementation, and will instead start goroutines that
  are responsible for routing the messages from the channel to libp2p
  fundamentals, replacing the current `p2p.Router.routeChannel`.
```

> - The current `p2p.Rotuer.dialPeers` and `p2p.Router.acceptPeers`,
>   which are responsible for establishing outbound and inbound
>   connections respectively, will be removed, along with
>   `p2p.Router.openConnection` and the libp2p connection manager will
>   be responsible for maintaining network connectivity.

Reviewer: ```suggestion
- The current `p2p.Router.dialPeers` and `p2p.Router.acceptPeers`,
  are responsible for establishing outbound and inbound connections, 
  respectively. These methods will be removed, along with
  `p2p.Router.openConnection`, and the libp2p connection manager will
  be responsible for maintaining network connectivity.
```

> - The `p2p.Channel` interface should change to replace the golang
>   channels with a more functional interface for sending messages. The
>   new methods on this object, will take contexts to support safe
>   cancellation and return errors, and generally block rather than
>   return asynchronously. The `Out` channel through which
>   reactors send messages to Peers, will be replaced by a `Send`
>   method, and the Error channel will be replaced by an `Error`
>   method. 

Reviewer: ```suggestion
- The `p2p.Channel` interface will change to replace Go
  channels with a more functional interface for sending messages.
  New methods on this object will take contexts to support safe
  cancellation, and return errors, and will block rather than
  running asynchronously. The `Out` channel through which
  reactors send messages to Peers, will be replaced by a `Send`
  method, and the Error channel will be replaced by an `Error`
  method. 
```

- Rectors will have access to an interface that will allow them to
  access information from the current state of the Peer information

>   from libp2p. This will supplant the  `p2p.PeerUpdates`

Reviewer 4: I think this is a great idea and I would also be intrigued as to what this interface would look like. AFAICT, knowing the height of a peer and the height of the chain are particularly important to reactor logic. There may be a few other things.

> - Rectors will have access to an interface that will allow them to
>   access information from the current state of the Peer information
>   from libp2p. This will supplant the  `p2p.PeerUpdates`
>   subscription.

Reviewer: ```suggestion
- Reactors will be passed an interface that will allow them to
  access Peer information from libp2p. This will supplant the
  `p2p.PeerUpdates` subscription.
```

### Upgrade and Compatibility

> Because the routers and all current P2P code is in the `internal`
> package and not part of the public API, the only changes to the public
> API surface area of the tendermint will be different configuration
> file options, replacing the current P2P options with options relevant
> to libp2p.

Reviewer: ```suggestion
Because the routers and all current P2P code is in `internal`
packages and not part of the public API, the only changes to the public
API surface area of Tendermint will be different configuration
file options, replacing the current P2P options with options relevant
to libp2p.
```

Maybe also we should plan to provide a tool to migrate existing settings to compatible equivalents in the new config section? Or maybe just agree to honor sensible interpretations of the old values for some period of time?

> However, it will not be possible to run a network with both networking

Reviewer 3: Yeah, I agree that given the circumstances, it's really not something we should spend too much time considering. Maybe we should reflect that context in this document in a few sentences? e.g.

```
Allow a network to run with both libp2p and the legacy stack will require maintaining both implementations within Tendermint simultaneously. Maintaining both solutions will inevitably be very complex and consume a lot of engineering time. Additionally, the need of a coordinated upgrade for other slated Tendermint features means there is already a natural point to switch from one to the other. We will therefore not allow networks to operate with both simultaneously.
```

> However, it will not be possible to run a network with both networking

Author: added a couple of sentences.

stacks active at once, so the upgrade to the version of Tendermint
will need to be coordinated between all nodes of the network.

## Open Questions

- What is the role of Protocol Labs in the implementation of libp2p in

>   tendermint, both during the initial implementation and on an ongoing

Reviewer: ```suggestion
  Tendermint, both during the initial implementation and on an ongoing
```

>   basis thereafter?

Reviewer 4: I second this. 

For me, I would want them to be active code reviewers to the implementation as well as offer a degree of guidance throughout the process. In the long term, I think it is important that we maintain channels of communication to offer both feedback and suggestions and to have some presence in their continual development

>   basis thereafter?

Reviewer 3: @tychoish this is the main point I think we should address before merging this adr.

>   basis thereafter?

Reviewer 5: > on other projects that are using libp2p

- Ethereum 2
- Polkadot
- Filecoin
- IPFS

Just to name a few. Happy to compile a longer list if need be.

> as well as information about how their roadmap may intersect/diverge from ours.

Assuming the above refers to libp2p and not the projects using libp2p: The libp2p roadmap can be found [here](https://github.com/libp2p/specs/blob/master/ROADMAP.md). I am not aware of any goal divergence.

> For me, I would want them to be active code reviewers to the implementation as well as offer a degree of guidance throughout the process.

Happy to review design documents and (in a limited extend) to be involved in code-reviews.

> I think it is important that we maintain channels of communication to offer both feedback and suggestions

We have a [public matrix channel](https://app.element.io/#/room/#libp2p:ipfs.io), a [forum](discuss.libp2p.io/) and the various GitHub repositories. Also feel free to reach out to me directly.

> and to have some presence in their continual development

That would be very much appreciated.

>   basis thereafter?

Reviewer 6: > on other projects that are using libp2p

a few more: 
- harmony
- [flow](https://github.com/onflow/flow-go) (also actively contributes to libp2p)
- mina (seem to have quite some issues but I don't know details)

- Should all P2P traffic for a given node be pushed to a single topic,
  and we assume that a topic maps to a specific ChainID, or should

>   each reactor (or type of message) have its own topic?

Reviewer: I'm not sure I understand the implications of the answer to this question. Previously, you mentioned topic-per-message as a possible strategy for obsoleting Router, but I thought that was mainly an implementation detail.

Is the question about whether the axis of variation for topics is (chanID x messageType) vs. just chainID vs. just messageType?

Do we have any information about how expensive it is to have a whole bunch of topics? It seems like libp2p wants topics to be lightweight and relatively inexpensive, but I don't have a good sense of the cost.

This ADR doesn't necessarily need to answer all that, but it may make sense to sketch out a few high-level criteria we can use for deciding once we know more.

Reviewer: https://docs.libp2p.io/concepts/publish-subscribe/ is probably a decent place to start. Broadly: A "topic" is a label publishers and subscribers use to coordinate. But I don't know how how expensive that abstraction is in libp2p.

Reviewer 2: Im a fan of per reactor topics. This would allow a form of service discovery in the network. Nodes that only run x reactor would only get subscribed to by peers looking for that reactor.

Author: I think the cost and scaling of the topic in libp2p is the like *core* question here, and I think we could go either way, if topics scale. 

As I've been thinking about it, I think I'm vaguely in the camp of one topic per logical consensus engine, but this might be a premature optimization for multi-chain, and because I think there's merit to retaining some amount of control over the message passing layer. 

This is definitely a thing we need PL feedback on.

> - Tendermint presently has a concept of message-type priority, which

Reviewer 2: QoS is needed for multiplexing a single tcp connection correct? If we use something like quic then another channel/stream wouldn't be able to starve another. This is how I understand quic.

Author: In a lot of ways the chief failing of the current system is that it pushes about 2/3rds of the QoS responsibility into the networking layer, which makes it really hard to do anything other than buffer a lot and hope. 

Since we're expecting to start off with TCP and iterate from there, I don't know that we can trust QUIC to solve this problem entirely. No matter how you abstract the socket (or groups of socket) the bits have to go over the same wire.

> - Tendermint presently has a concept of message-type priority, which
>   provides a very course QoS-like functionality and
>   intuitively/theoretically ensures that evidence and consensus
>   messages don't get starved by blocksync/statesync messages. It's
>   unclear if we should attempt to replicate this with libp2p and even
>   if we should.

Reviewer: ```suggestion
- Tendermint presently provides a very coarse QoS-like functionality 
  using priorities based on message-type.
  This intuitively/theoretically ensures that evidence and consensus
  messages don't get starved by blocksync/statesync messages. It's
  unclear if we should attempt to replicate this with libp2p and even
  if we should.
```

Having _some_ means of talking about QoS almost certainly makes sense. Doing it by type is probably not the right abstraction (though it was a clever trick for the old system).

Reviewer 4: Are we aware of what current mechanisms libp2p have in the way of QoS? My second question would be how do we gauge the effectiveness of QoS in a p2p layer?

Author: > Are we aware of what current mechanisms libp2p have in the way of QoS? My second question would be how do we gauge the effectiveness of QoS in a p2p layer?

Yeah, this is really at the heart of what we need to talk about with PL

> - Is it possible to attach additional (and potentially arbitrary)
>   information into the DHT as part of the heartbeats between nodes,
>   such as the latest height, and then access that in arbitrary
>   reactors.

Reviewer: ```suggestion
- Is it possible to store additional (potentially arbitrary)
  data into the DHT as part of the heartbeats between nodes,
  such as the latest height, and then access that in the
  reactors.
```

This is a neat idea, though I imagine we would have to be very careful about the security implications (e.g., encrypting and/or signing such data, since the DHT is not protected by the BFT resistance of consensus, so anything we put there needs some basis for trust).

Reviewer 2: huge fan of this! it would help organise networks and peers. Most likely reduce amount hops when searching for things.

Reviewer 4: This could relate well with the peer info work mentioned above

Reviewer 4: I also share the same concerns as Michael in regards to the security implications of this. I would recommend starting with a small prescribed set of data yet allow for future additions. I don't think we should allow the application to have some input on this just yet but it could be something we consider in the future.

Reviewer 4: Also what are the characteristics of DHT's. Do they work well with lots of writes. Does it make sense for writes by each node at each height?

Reviewer 4: I was talking to someone from Celestia yesterday who had used DHT's in LibP2P and said it wasn't designed for frequent writes and believed it should just be used for peer discovery and potentially service discovery (as we don't expect nodes to change their services too frequently). For the frequent broadcasting of arbitrary data like latest height we should probably use some other mechanism (might even be a small reactor)

  
- Does it make sense to have reactors continue to consume inbound
  messages from a Channel (`In`) or is there another interface or

>   pattern that we should consider?

Reviewer: A `Recv` analogue of the proposed `Send` method could work well. The reactor could still fan out work in parallel internally if it wants, and in fact a naive implementation could be built around the existing channels.

Reviewer 3: The difficulty with removing channels is that, to work alongside a context, you likely have to have a channel internally anyway so that you can select against `context.Done()`. I somewhat prefer to using a channel in the p2p API so that we can make the contract a little bit more clear. A `Recv() (Value, Error)` method seems pretty clean. It enables a simple loop, while making it clear that the caller has no responsibility to 'close' the `Recv` beyond, perhaps closing the object provided by the p2p code.

Reviewer: I don't follow that entirely.  I think it's orthogonal: If the sender and receiver rendezvous, whether or not there is a channel under the covers, they have to agree on what it means to be "finished". If the channels are exposed in the API, then the client has to cooperate to make that happen, but if not, then it can be handled transparently.

Either way, though, we presumably want a context for lifecycle management. It's just that in the case where we hide the channels, the context is the ONLY visible mechanism we really need. But maybe I missed some aspect of your concern?

Reviewer: Oh, I think I understand now. I misread what you wrote as advocating _for_ a channel in the API, rather than against. I think we are actually in complete agreement.

Reviewer 3: Sorry!  I accidentally a word.

Reviewer 4: Also agree

Author: I don't think that `<-chan` typed channels are super different from `Recv()` methods, though again, I'm open to both ends, and deeply in the camp of "get rid of as many channels as possible.

I think from a code perspective, however, keeping `In` as a Channel rather than replacing it with an `Recv` method makes it easy to write handler functions that consume and process messages from many different channels (e.g. as in the consensus and statesync reactors where we're consuming a number of channels) It's tractable, but that means having a processing goroutine for every channel, which just makes the locking story within the reactor implementation a bit more dodgy.

Reviewer: Re: Many input channels. Without reflection, we can only `select` on a static set of inputs, not an arbitrary collection. If we really have cases where, for example, it helps to read exactly k channels at once, I certainly don't oppose using channels, but I haven't really seen an example of that in our code. I think "not exposing channels in the API" is a good default position, that it's fine to deviate from iff we have a concrete need to do so.

> 

Reviewer 2: Another open question may be, how much control do we want to give to the user. Libp2p has nice interfaces for swapping network designs, kademlia, pex, Gnutella, etc.. Many networks would probably not come this low in the stack but the ones that are trying to scale may want to use a different protocol then we default to.

Reviewer: IMO the right answer, at least at first, is "expose as little configuration to the user as possible". It's easier to add more knobs than it is to ever retract them.

## Consequences

### Positive

- Reduce the maintenance burden for the Tendermint Core team by
  removing a large swath of legacy code that has proven to be

>   difficult to modify safely.

Reviewer 4: Not just legacy code but the entire burden of maintaining and developing a p2p stack.

- Provide users with a more stable peer and networking system,
  Tendermint can improve operator experience and network stability.

### Negative

- By deferring to library implementations for peer management and
  networking, Tendermint loses some flexibility for innovating at the
  peer and networking level. However, Tendermint should be innovating
  primarily at the consensus layer, and libp2p does not preclude
  optimization or development in the peer layer.

- Libp2p is a large dependency and Tendermint would become dependent
  upon Protocol Labs' release cycle and prioritization for bug
  fixes. If this proves onerous, it's possible to maintain a vendor
  fork of relevant components as needed.

### Neutral

- N/A

## References

> - [ADR 61: P2P Refactor Scope](./adr-061-p2p-refactor-scope.md)
> - [ADR 62: P2P Architecture](./adr-062-p2p-architecture.md)
> - [P2P Roadmap RFC](../rfc/rfc-000-p2p.rst)

Reviewer: ```suggestion
- [ADR 61: P2P Refactor Scope][adr61]
- [ADR 62: P2P Architecture][adr62]
- [P2P Roadmap RFC][rfc]

[adr61]: ./adr-061-p2p-refactor-scope.md
[adr62]: ./adr-062-p2p-architecture.md
[rfc]: ../rfc/rfc-000-p2p.rst
```

# ADR 073: Adopt LibP2P

## Changelog

- 2021-11-02: Initial Draft (@tychoish)

## Status

Proposed.

## Context

As part of the 0.35 development cycle, the Tendermint team completed
the first phase of the work described in ADRs 61 and 62, which included a
large scale refactoring of the reactors and the p2p message
routing. This replaced the switch and many of the other legacy

> components, without breaking protocol or network-level

Reviewer 4: ```suggestion
components without breaking protocol or network-level
```

interoperability and left the legacy connection/socket handling code.

Following the release, the team has reexamined the state of the code
and the design, as well as Tendermint's requirements. The notes
from that process are available in the [P2P Roadmap
RFC][rfc].

This ADR supersedes the decisions made in ADRs 60 and 61, but
builds on the completed portions of this work. Previously, the
boundaries of peer management, message handling, and the higher level
business logic (e.g., "the reactors") were intermingled, and core
elements of the p2p system were responsible for the orchestration of
higher-level business logic. Refactoring the legacy components
made it more obvious that this entanglement of responsibilities
had outsized influence on the entire implementation, making
it difficult to iterate within the current abstractions.
It would not be viable to maintain interoperability with legacy
systems while also achieving many of our broader objectives.

LibP2P is a thoroughly-specified implementation of a peer-to-peer
networking stack, designed specifically for systems such as

> ours. Adopting LibP2P as the basis of the Tendermint will allow the

Reviewer 4: ```suggestion
ours. Adopting LibP2P as the basis of Tendermint will allow the
```

level connection and peer management code make this operation
relatively straightforward to implement. A key
goal in this design is to minimize the impact on the reactors
(potentially entirely,) and completely remove the lower level
components (e.g., `Transport`, `Connection` and `PeerManager`) using the
separation afforded by the `Router` layer. The current state of the
code makes these changes relatively surgical, and limited to a small
number of methods:

- `p2p.Router.OpenChannel` will still return a `Channel` structure
  which will continue to serve as a pipe between the reactors and the
  `Router`. The implementation will no longer need the queue
  implementation, and will instead start goroutines that
  are responsible for routing the messages from the channel to libp2p
  fundamentals, replacing the current `p2p.Router.routeChannel`.

- The current `p2p.Router.dialPeers` and `p2p.Router.acceptPeers`,
  are responsible for establishing outbound and inbound connections,
  respectively. These methods will be removed, along with
  `p2p.Router.openConnection`, and the libp2p connection manager will
  be responsible for maintaining network connectivity.

> - The `p2p.Channel` interface will change to replace Go

Reviewer 2: this may be default in libp2p but do you know if Tendermint will multiplex the connection like we are doing now, or if that needed?

  channels with a more functional interface for sending messages.
  New methods on this object will take contexts to support safe
  cancellation, and return errors, and will block rather than
  running asynchronously. The `Out` channel through which
  reactors send messages to Peers, will be replaced by a `Send`
  method, and the Error channel will be replaced by an `Error`
  method.

- Reactors will be passed an interface that will allow them to
  access Peer information from libp2p. This will supplant the

>   `p2p.PeerUpdates` subscription.

Reviewer 4: Nice 👍

the implementation timeline.

## Open Questions

- What is the role of Protocol Labs in the implementation of libp2p in
  tendermint, both during the initial implementation and on an ongoing
  basis thereafter?

- Should all P2P traffic for a given node be pushed to a single topic,
  so that a topic maps to a specific ChainID, or should
  each reactor (or type of message) have its own topic? How many
  topics can a libp2p network support? Is there testing that validates
  the capabilities?

- Tendermint presently provides a very coarse QoS-like functionality
  using priorities based on message-type.
  This intuitively/theoretically ensures that evidence and consensus
  messages don't get starved by blocksync/statesync messages. It's
  unclear if we can or should attempt to replicate this with libp2p.

- What kind of QoS functionality does libp2p provide and what kind of
  metrics does libp2p provide about it's QoS functionality?

- Is it possible to store additional (and potentially arbitrary)
  information into the DHT as part of the heartbeats between nodes,
  such as the latest height, and then access that in the

>   reactors. How frequently can the DHT be updated?

Reviewer 5: Yes you can store arbitrary data in the libp2p Kademlia DHT. To judge whether that is a good idea, I would need more context.

The seams in the P2P implementation between the higher level
constructs (reactors), the routing layer (`Router`) and the lower
level connection and peer management code make this operation
relatively straightforward to implement. A key
goal in this design is to minimize the impact on the reactors
(potentially entirely,) and completely remove the lower level
components (e.g., `Transport`, `Connection` and `PeerManager`) using the
separation afforded by the `*Router` layer. The current state of the
code makes these changes relatively surgical, and limited to a small
number of methods:

- `p2p.Router.OpenChannel` will still return a `Channel` structure
  which will continue to serve as a pipe between the reactors and the
  `Router`. The implementation will no longer need the queue
  implementation, and will instead start goroutines that
  are responsible for routing the messages from the channel to libp2p
  fundamentals, replacing the current `p2p.Router.routeChannel`.

- The current `p2p.Router.dialPeers` and `p2p.Router.acceptPeers`,
  are responsible for establishing outbound and inbound connections,
  respectively. These methods will be removed, along with

>   `p2p.Router.openConnection`, and the libp2p connection manager will

Reviewer: A question I've seen come up elsewhere is whether other parts of the system need to be made aware when connections are made and lost, and whether libp2p provides a way for us to address that concern.

Even if we don't have a complete answer yet, it might be worth mentioning that issue (if it is real).

Author: well, maybe? I think this is a thing that we presently do a lot of and spend time keeping reactors notified of the peer network state, but I don't know that reactors always (or even often) need to know this information, and I think often it's for premature optimization reasons.

I think it's possible that a node could deduce this information from other messages or take advantage of information in heartbeats or the DHT to answer some of these questions. 

It's good and important to call out though.

  be responsible for maintaining network connectivity.

- The `p2p.Channel` interface will change to replace Go
  channels with a more functional interface for sending messages.
  New methods on this object will take contexts to support safe
  cancellation, and return errors, and will block rather than
  running asynchronously. The `Out` channel through which
  reactors send messages to Peers, will be replaced by a `Send`
  method, and the Error channel will be replaced by an `Error`
  method.

- Reactors will be passed an interface that will allow them to
  access Peer information from libp2p. This will supplant the
  `p2p.PeerUpdates` subscription.

- Add some kind of heartbeat message at the application level
  (e.g. with a reactor,) potentially connected to libp2p's DHT to be
  used by reactors for service discovery, message targeting, or other
  features.

> - Replace the existing/legacy handshake protocol with Noise.

Reviewer: ```suggestion
- Replace the existing/legacy handshake protocol with [Noise](http://www.noiseprotocol.org/noise.html).
```

> By default, this project will take advantage of TCP-based transport
> protocols within libp2p, but QUIC will be available as an
> option. Mixed networks are unlikely to be supported, certainly in the
> initial release, but perhaps at any point.

Reviewer: I suggest being more opinionated here:

```suggestion
This project will initially use the TCP-based transport protocols within
libp2p. QUIC is also available as an option that we may implement later.
We will not support mixed networks in the initial release, but will
revisit that possibility later if there is a demonstrated need.
```

### Upgrade and Compatibility

> Because the routers and all current P2P code is in `internal`
> packages and not part of the public API, the only changes to the public
> API surface area of Tendermint will be different configuration
> file options, replacing the current P2P options with options relevant
> to libp2p.

Reviewer: Completely cosmetic number agreement nit.
```suggestion
Because the routers and all current P2P libraries are `internal`
packages and not part of the public API, the only changes to the public
API surface area of Tendermint will be different configuration
file options, replacing the current P2P options with options relevant
to libp2p.
```

However, it will not be possible to run a network with both networking
stacks active at once, so the upgrade to the version of Tendermint
will need to be coordinated between all nodes of the network. This is
consistent with the expectations around upgrades for Tendermint moving
forward, and will help manage both the complexity of the project and
the implementation timeline.

## Open Questions

- What is the role of Protocol Labs in the implementation of libp2p in
  tendermint, both during the initial implementation and on an ongoing
  basis thereafter?

> - Should all P2P traffic for a given node be pushed to a single topic,
>   and we assume that a topic maps to a specific ChainID, or should
>   each reactor (or type of message) have its own topic? How many
>   topics can a libp2p network support? Is there testing that validates
>   the capabilities?

Reviewer: ```suggestion
- Should all P2P traffic for a given node be pushed to a single topic,
  so that a topic maps to a specific ChainID, or should
  each reactor (or type of message) have its own topic? How many
  topics can a libp2p network support? Is there testing that validates
  the capabilities?
```

- Tendermint presently provides a very coarse QoS-like functionality
  using priorities based on message-type.
  This intuitively/theoretically ensures that evidence and consensus

>   messages don't get starved by blocksync/statesync messages. It's
>   unclear if we should attempt to replicate this with libp2p and even
>   if we should.

Reviewer: ```suggestion
  messages don't get starved by blocksync/statesync messages. It's
  unclear if we can or should attempt to replicate this with libp2p.
```

- What kind of QoS functionality does libp2p provide and what kind of
  metrics does libp2p provide about it's QoS functionality?

- Is it possible to store additional (and potentially arbitrary)
  information into the DHT as part of the heartbeats between nodes,
  such as the latest height, and then access that in the
  reactors. How frequently can the DHT be updated?

- Does it make sense to have reactors continue to consume inbound
  messages from a Channel (`In`) or is there another interface or
  pattern that we should consider?

  - We should avoid exposing Go channels when possible, and likely
	some kind of alternate iterator likely makes sense for processing
	messages within the reactors.

- What are the security and protocol implications of tracking
  information from peer heartbeats and exposing that to reactors?

- How much (or how little) configuration can Tendermint provide for
  libp2p, particularly on the first release?

  - In general, we should not support byo-functionality for libp2p
	components within Tendermint, and reduce the configuration surface
	area, as much as possible.

> - What are the best ways to provide request/response semantics for

Reviewer: Do the reactors depend on request/response semantics now, or is that something we want for future work?

Author: they currently have a request/response workload, and operate that way (e.g. sending messages to specific peers and handling incoming messages from those peers) but the responses aren't connected to the requests and the system doesn't do any sort of  tracking to make sure that requests are fulfilled before sending more requests or any other kind of tracking.

Reviewer: Ah, ok. I've seen places in the code where it does as you describe, but I thought that was more about tracking peers than specific requests. Intuitively, I think we can easily do at least that well (e.g., a PoC would be to attach UUIDs to messages that we need to correlate, so even if that's not the best solution, a solution exists).

Author: Yeah, the UUID thing was what I was planning to do because it's intelligible/obvious.

Reviewer 4: Currently, each reactor: `statesync`, `blocksync`, `pex` (and in the future the light client), all implement their own request response mechanism. I think in all the cases, the request is tracked by the `NodeID` (thus you can't make two simultaneous requests to the same peer). Each reactor keeps their own local list of peers that they update using the `PeerUpdates` channel. In most cases they just loop through this list when fetching data. It would be good to have a single request response mechanism that reactors can use instead of having multiple implementations.
